src/main/java/org/opensearch/security/auth/UserInjector.java

```diff
@@ -123,59 +87,74 @@ public InjectedUser getInjectedUser() {
         }
 
         // username
-        if (Strings.isNullOrEmpty(parts[0])) {
+        String userName = parts[0];
+        if (Strings.isNullOrEmpty(userName)) {
             log.error("Username must not be null, user string was '{}.' User injection failed.", injectedUserString);
             return null;
         }
 
-        final InjectedUser injectedUser = new InjectedUser(parts[0]);
-
         // backend roles
+        ImmutableSet<String> backendRoles = ImmutableSet.of();
         if (parts.length > 1 && !Strings.isNullOrEmpty(parts[1])) {
             if (parts[1].length() > 0) {
-                injectedUser.addRoles(Arrays.asList(parts[1].split(",")));
+                backendRoles = ImmutableSet.copyOf(Arrays.asList(parts[1].split(",")));
             }
         }
 
         // custom attributes
+        ImmutableMap<String, String> customAttributes = ImmutableMap.of();
         if (parts.length > 3 && !Strings.isNullOrEmpty(parts[3])) {
-            Map<String, String> attributes = mapFromArray((parts[3].split(",")));
-            if (attributes == null) {
+            customAttributes = mapFromArray((parts[3].split(",")));
+            if (customAttributes == null) {
                 log.error("Could not parse custom attributes {}, user injection failed.", parts[3]);
                 return null;
-            } else {
-                injectedUser.addAttributes(attributes);
             }
         }
 
         // requested tenant
+        String requestedTenant = null;
         if (parts.length > 4 && !Strings.isNullOrEmpty(parts[4])) {
-            injectedUser.setRequestedTenant(parts[4]);
+            requestedTenant = parts[4];
         }
 
         // remote IP - we can set it only once, so we do it last. If non is given,
         // BackendRegistry/XFFResolver will do the job
+        TransportAddress transportAddress = null;
         if (parts.length > 2 && !Strings.isNullOrEmpty(parts[2])) {
             try {
-                injectedUser.setTransportAddress(parts[2]);
+                transportAddress = parseTransportAddress(parts[2]);
             } catch (UnknownHostException | IllegalArgumentException e) {
                 log.error("Cannot parse remote IP or port: {}, user injection failed.", parts[2], e);
                 return null;
             }
         }
 
-        // mark user injected for proper admin handling
-        injectedUser.setInjected(true);
+        User injectedUser = new User(userName, backendRoles, ImmutableSet.of(), requestedTenant, customAttributes, true);
 
         if (log.isTraceEnabled()) {
             log.trace("Injected user object:{} ", injectedUser.toString());
         }
 
-        return injectedUser;
+        return new Result(injectedUser, transportAddress);
+    }
+
+    public TransportAddress parseTransportAddress(String addr) throws UnknownHostException, IllegalArgumentException {
```

Comment:
FYI: This method is not new, it was just moved from the now deleted inner class `InjectedUser`